common/archiver/s3store/historyArchiver.go

```diff
@@ -223,38 +267,61 @@ func (h *historyArchiver) Get(
 		}
 		token = &getHistoryToken{
 			CloseFailoverVersion: *highestVersion,
-			NextBatchIdx:         0,
 		}
 	}
 
-	key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion)
-	encodedHistoryBatches, err := download(ctx, h.s3cli, URI, key)
+	prefix := constructHistoryKeyPrefixWithVersion(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion)
+
+	results, err := h.s3cli.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket:     aws.String(URI.Hostname()),
+		Prefix:     aws.String(prefix),
+		MaxKeys:    aws.Int64(int64(request.PageSize)), // Upper bound of history events
+		StartAfter: (*token).StartAfterKey,
```

Comment:
`token.StartAfterKey` should be enough?

Just to confirm: the key specified here is exclusive right?

Reply:
Yes, the key is exclusive